op-conductor/rpc/ws/server.go

```diff
@@ -0,0 +1,270 @@
+package ws
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/coder/websocket"
+	"github.com/ethereum-optimism/optimism/op-conductor/metrics"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// Hub maintains the set of active clients and broadcasts messages to them
+type Hub struct {
+	// Registered clients
+	clients map[*Client]bool
+
+	// Register requests from the clients
+	register chan *Client
+
+	// Unregister requests from clients
+	unregister chan *Client
+
+	// Inbound messages to broadcast to the clients
+	broadcast chan []byte
+
+	// Signal to stop the hub
+	done chan struct{}
+
+	// Logger
+	log log.Logger
+
+	// Metrics
+	metrics metrics.Metricer
+
+	// Callbacks for testing and monitoring
+	callbacks HubCallbacks
+}
+
+// newHub creates a new hub
+func newHub(m metrics.Metricer) *Hub {
+	return &Hub{
+		broadcast:  make(chan []byte, 256),
+		register:   make(chan *Client),
+		unregister: make(chan *Client),
+		clients:    make(map[*Client]bool),
+		done:       make(chan struct{}),
+		log:        log.New("component", "websocket-hub"),
+		metrics:    m,
+	}
+}
+
+// registerClient adds a client to the hub and updates metrics
+func (h *Hub) registerClient(client *Client) {
+	h.clients[client] = true
+	clientCount := len(h.clients)
+	h.log.Info("Client registered with hub", "totalClients", clientCount)
+	h.metrics.RecordWebSocketClientCount(clientCount)
+
+	if h.callbacks.OnClientRegistered != nil {
+		h.callbacks.OnClientRegistered(client)
+	}
+}
+
+// unregisterClient removes a client from the hub, closes it, and updates metrics
+func (h *Hub) unregisterClient(client *Client) {
+	if _, ok := h.clients[client]; ok {
+		delete(h.clients, client)
+		client.Close()
+		clientCount := len(h.clients)
+		h.log.Info("Client unregistered from hub", "totalClients", clientCount)
+		h.metrics.RecordWebSocketClientCount(clientCount)
```

Comment:
Note for future PR:
- `m.webSocketClients.Dec()` can be leveraged here, removing the need to find-and-set the gauge value.